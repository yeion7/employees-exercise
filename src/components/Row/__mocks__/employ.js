const employ = {
  id: 0,
  name: 'Eduardo González',
  company: 'América Móvil',
  salary: 5400,
  age: 57,
  phone: '(886) 271 4211',
  email: 'eduardo@gmail.com'
}

export default employ
