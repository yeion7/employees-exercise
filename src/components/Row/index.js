import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

/**
 * This element use PureComponent for optimize edit mode
 * If edit a field not re-render every field
 */

class Row extends PureComponent {
  render() {
    const {
      id,
      name,
      company,
      salary = 0,
      age,
      phone,
      email,
      added,
      currency,
      changeType,
      deleteEmploy,
      editable,
      onEditField
    } = this.props

    const salaryValue = currency === 'USD' ? salary / changeType : salary

    return (
      <tr>
        <td>
          {editable ? (
            <input
              data-test={`input-name-${id}`}
              defaultValue={name}
              onChange={onEditField(id, 'name')}
            />
          ) : (
            <span>{name}</span>
          )}
        </td>
        <td>
          {editable && added ? (
            <input
              data-test={`input-company-${id}`}
              defaultValue={company}
              onChange={onEditField(id, 'company')}
            />
          ) : (
            <span>{company}</span>
          )}
        </td>
        <td>
          {editable ? (
            <input
              data-test={`input-salary-${id}`}
              type="number"
              defaultValue={salary}
              onChange={onEditField(id, 'salary')}
            />
          ) : (
            <span
              style={{ color: `${salary < 10000 ? 'indianred' : 'seagreen'}` }}
            >
              {parseInt(salaryValue, 10).toLocaleString('es-MX', {
                style: 'currency',
                currency: currency,
                minimumFractionDigits: 2,
                maximumFractionDigits: 2
              })}
            </span>
          )}
        </td>
        <td>
          {editable ? (
            <input
              data-test={`input-age-${id}`}
              type="number"
              defaultValue={age}
              onChange={onEditField(id, 'age')}
            />
          ) : (
            <span>{age}</span>
          )}
        </td>
        <td>
          {editable ? (
            <input
              data-test={`input-phone-${id}`}
              defaultValue={phone}
              onChange={onEditField(id, 'phone')}
            />
          ) : (
            <span>{phone}</span>
          )}
        </td>
        <td>
          {editable ? (
            <input
              data-test={`input-email-${id}`}
              type="email"
              defaultValue={email}
              onChange={onEditField(id, 'email')}
            />
          ) : (
            <span>{email}</span>
          )}
        </td>
        <td>
          <button data-test={`delete-${id}`} onClick={() => deleteEmploy(id)}>
            Borrar
          </button>
        </td>
      </tr>
    )
  }
}

Row.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  name: PropTypes.string.isRequired,
  company: PropTypes.string.isRequired,
  salary: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  age: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  phone: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired
}

export default Row
