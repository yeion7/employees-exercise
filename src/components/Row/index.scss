.row {
  &:nth-child(even) {
    background: aliceblue;
  }

  &:hover {
    background: #f4b642;
  }

  &__cell {
    padding: 10px 20px;

    &--right {
      text-align: right;
      font-family: monospace, monospace;
    }
  }

  &__button {
    cursor: pointer;
    font-size: 1.2em;
    padding: 5px;
    background: transparent;
    border: 0;
  }
}
