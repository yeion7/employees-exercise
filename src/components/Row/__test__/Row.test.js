import React from 'react'
import renderer from 'react-test-renderer'

import Row from '../'
import employ from '../__mocks__/employ'

describe('Row component', () => {
  it('renders correctly with MXN', () => {
    const fn = jest.fn()

    const tree = renderer
      .create(
        <Row
          id={employ.id}
          name={employ.name}
          company={employ.company}
          salary={employ.salary}
          age={employ.age}
          phone={employ.phone}
          email={employ.email}
          currency='MXN'
          changeType={19.9}
          editable={false}
          deleteEmploy={fn}
          onEditField={fn}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with USD', () => {
    const fn = jest.fn()

    const tree = renderer
      .create(
        <Row
          id={employ.id}
          name={employ.name}
          company={employ.company}
          salary={employ.salary}
          age={employ.age}
          phone={employ.phone}
          email={employ.email}
          currency='USD'
          changeType={19.9}
          editable={false}
          deleteEmploy={fn}
          onEditField={fn}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with editable mode', () => {
    const fn = jest.fn()
    const factory = jest.fn(() => jest.fn())

    const tree = renderer
      .create(
        <Row
          id={employ.id}
          name={employ.name}
          company={employ.company}
          salary={employ.salary}
          age={employ.age}
          phone={employ.phone}
          email={employ.email}
          currency='USD'
          changeType={19.9}
          editable
          deleteEmploy={fn}
          onEditField={factory}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly without salary', () => {
    const fn = jest.fn()
    const factory = jest.fn(() => jest.fn())

    const tree = renderer
      .create(
        <Row
          id={employ.id}
          name={employ.name}
          company={employ.company}
          age={employ.age}
          phone={employ.phone}
          email={employ.email}
          currency='USD'
          changeType={19.9}
          editable
          deleteEmploy={fn}
          onEditField={factory}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
