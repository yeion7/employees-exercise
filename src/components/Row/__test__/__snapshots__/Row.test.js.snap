// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row component renders correctly with MXN 1`] = `
<tr>
  <td>
    <span>
      Eduardo González
    </span>
  </td>
  <td>
    <span>
      América Móvil
    </span>
  </td>
  <td>
    <span
      style={
        Object {
          "color": "indianred",
        }
      }
    >
      MX$ 5,400.00
    </span>
  </td>
  <td>
    <span>
      57
    </span>
  </td>
  <td>
    <span>
      (886) 271 4211
    </span>
  </td>
  <td>
    <span>
      eduardo@gmail.com
    </span>
  </td>
  <td>
    <button
      data-test="delete-0"
      onClick={[Function]}
    >
      Borrar
    </button>
  </td>
</tr>
`;

exports[`Row component renders correctly with USD 1`] = `
<tr>
  <td>
    <span>
      Eduardo González
    </span>
  </td>
  <td>
    <span>
      América Móvil
    </span>
  </td>
  <td>
    <span
      style={
        Object {
          "color": "indianred",
        }
      }
    >
      US$ 271.00
    </span>
  </td>
  <td>
    <span>
      57
    </span>
  </td>
  <td>
    <span>
      (886) 271 4211
    </span>
  </td>
  <td>
    <span>
      eduardo@gmail.com
    </span>
  </td>
  <td>
    <button
      data-test="delete-0"
      onClick={[Function]}
    >
      Borrar
    </button>
  </td>
</tr>
`;

exports[`Row component renders correctly with editable mode 1`] = `
<tr>
  <td>
    <input
      data-test="input-name-0"
      defaultValue="Eduardo González"
      onChange={[Function]}
    />
  </td>
  <td>
    <span>
      América Móvil
    </span>
  </td>
  <td>
    <input
      data-test="input-salary-0"
      defaultValue={5400}
      onChange={[Function]}
      type="number"
    />
  </td>
  <td>
    <input
      data-test="input-age-0"
      defaultValue={57}
      onChange={[Function]}
      type="number"
    />
  </td>
  <td>
    <input
      data-test="input-phone-0"
      defaultValue="(886) 271 4211"
      onChange={[Function]}
    />
  </td>
  <td>
    <input
      data-test="input-email-0"
      defaultValue="eduardo@gmail.com"
      onChange={[Function]}
      type="email"
    />
  </td>
  <td>
    <button
      data-test="delete-0"
      onClick={[Function]}
    >
      Borrar
    </button>
  </td>
</tr>
`;

exports[`Row component renders correctly without salary 1`] = `
<tr>
  <td>
    <input
      data-test="input-name-0"
      defaultValue="Eduardo González"
      onChange={[Function]}
    />
  </td>
  <td>
    <span>
      América Móvil
    </span>
  </td>
  <td>
    <input
      data-test="input-salary-0"
      defaultValue={0}
      onChange={[Function]}
      type="number"
    />
  </td>
  <td>
    <input
      data-test="input-age-0"
      defaultValue={57}
      onChange={[Function]}
      type="number"
    />
  </td>
  <td>
    <input
      data-test="input-phone-0"
      defaultValue="(886) 271 4211"
      onChange={[Function]}
    />
  </td>
  <td>
    <input
      data-test="input-email-0"
      defaultValue="eduardo@gmail.com"
      onChange={[Function]}
      type="email"
    />
  </td>
  <td>
    <button
      data-test="delete-0"
      onClick={[Function]}
    >
      Borrar
    </button>
  </td>
</tr>
`;
